### ISSUE TITLE
Integer variable assignment in for loop different between Cuda and CPU

### ISSUE THREAD

**Issue by rhoadscm** [m4rs-mt/ILGPU#700]
I was implementing a ray casting point-in-polygon algorithm and needed to iterate over a number of vertexes within a kernel while keeping track of the previous vertex and discovered that variable assignment for `int` and `long` seemed to behave differently between the GPU and CPU. Here is the kernel code to reproduce it using the latest nuget package release and an NVIDIA Quadro M1200:

```cs
static void WriteLineKernel(Index1 index, ArrayView<int> dataView)
{
    long i, j = dataView.Length - 1;
    for(i = 0; i < dataView.Length; i++)
    {
        Interop.WriteLine("{0}, {1}", i, j);
        j = i;
    }
}
```
```
output:
0, 4
1, 1
2, 2
3, 3
4, 4
```

This was launched on a single thread to illustrate the problem. The expectation is that `j` lags `i` by 1 on each iteration, but the output shows they are identical for every iteration except the first. Switching from a CudaAccelerator to a CPUAccelerator and running the exact same kernel produces:

```
output:
0, 4
1, 0
2, 1
3, 2
4, 3
```
Which is what I expected.

One other thing to note is that changing `for(i = 0; i < dataView.Length; i++)` to `for(i = 0; i < 5; i++)` will produce the same output with both the Cuda and CPU accelerators.

Full Program.cs:
```cs
using ILGPU;
using ILGPU.Runtime;
using ILGPU.Runtime.CPU;
using ILGPU.Runtime.Cuda;
using System.Linq;

namespace ILGPU_Test
{
    class Program
    {
        static void WriteLineKernel(Index1 index, ArrayView<int> dataView)
        {
            long i, j = dataView.Length - 1;
            for(i = 0; i < 5; i++)
            {
                Interop.WriteLine("{0}, {1}", i, j);
                j = i;
            }
        }

        static void Main()
        {
            var values = Enumerable.Range(0, 5).ToArray();

            using Context context = new();

            using var accelerator = new CudaAccelerator(context);

            var kernel = accelerator.LoadAutoGroupedStreamKernel<Index1, ArrayView<int>>(WriteLineKernel);
            using var buffer = accelerator.Allocate(values);

            kernel(1, buffer.View);
            accelerator.DefaultStream.Synchronize();
            
        }
    }
}
```

**Comment by rhoadscm** [comment]
My workaround is this:
```cs
static void WriteLineKernel(Index1 index, ArrayView<int> dataView)
{
    long i, j;
    for(i = 0; i < dataView.Length; i++)
    {
        if (i == 0)
            j = dataView.Length - 1;
        else
            j = i - 1;
        Interop.WriteLine("{0}, {1}", i, j);
    }
}
```
It works fine, but it introduces and extra operation for every iteration.

**Comment by m4rs-mt** [comment]
@rhoadscm Thank you very much for your detailed issue report. I can reproduce this issue using `v0.10` stable. However, this issue does neither occur in any of the `v1.0.0-rcX` release candidates nor in the `v1.0.0` stable (according to my tests). It should have been fixed in #541.

**Comment by rhoadscm** [comment]
Thanks for the reply and the fix, much appreciated!
